manta-trusted-setup/src/groth16/mpc.rs

```diff
@@ -54,14 +61,54 @@ pub struct State<P>(
         feature = "serde",
         serde(
             serialize_with = "canonical_serialize::<ProvingKey<P::Pairing>, _>",
-            deserialize_with = "canonical_deserialize::<'de, _, ProvingKey<P::Pairing>>"
+            deserialize_with = "canonical_deserialize_unchecked::<'de, _, ProvingKey<P::Pairing>>"
         )
     )]
     pub ProvingKey<P::Pairing>,
 )
 where
     P: Pairing + ?Sized;
 
+// TODO: A macro should impl this for SWModel and TEModel
+impl<P, R1, R2> State<P>
+where
+    P: Pairing<G1 = GroupAffine<R1>, G2 = GroupAffine<R2>> + ?Sized,
+    R1: SWModelParameters,
+    R2: SWModelParameters,
+{
+    /// We deserialize a [`State`] unchecked by default, so a ceremony
+    /// coordinator needs to perform the deserialization checks manually.
+    #[inline]
+    pub fn check(&self) -> Result<(), SerializationError> {
+        once(&self.0.vk.alpha_g1)
+            .chain(self.0.vk.gamma_abc_g1.iter())
+            .chain(once(&self.0.beta_g1))
+            .chain(once(&self.0.delta_g1))
+            .chain(self.0.a_query.iter())
+            .chain(self.0.b_g1_query.iter())
+            .chain(self.0.h_query.iter())
+            .chain(self.0.l_query.iter())
+            .try_for_each(curve_point_checks)?;
+
+        once(&self.0.vk.beta_g2)
+            .chain(once(&self.0.vk.gamma_g2))
+            .chain(once(&self.0.vk.delta_g2))
+            .chain(self.0.b_g2_query.iter())
+            .try_for_each(curve_point_checks)
+    }
+}
+
+fn curve_point_checks<P>(p: &GroupAffine<P>) -> Result<(), SerializationError>
```

Comment:
Add documentation and `#[inline]` marker, and probably move to `manta_crypto::arkworks::serialize`.